---
title: "Pan-Gene Wide Association Study of "
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

##Load Lsibraries and Download Data

```{r}
library(data.table)
library(ggplot2)
library(knitr)
library(purrr)
library(stringr)
library(piggyback)
```

```{r, eval=FALSE}
pb_download("gono_harris_gwas.zip", 
            repo = "gtonkinhill/panaroo_manuscript", 
            dest = "./")
unzip("gono_harris_gwas.zip")
```

To demonstrate the use of Panaroo with Pyseer to preform association analyses between gene presence/absence, structural variations and phenotypes we investigated the *Neisseria gonorrhoeae* dataset of:

```
Harris,S.R., Cole,M.J., Spiteri,G., Sánchez-Busó,L., Golparian,D., Jacobsson,S., Goater,R., Abudahab,K., Yeats,C.A., Bercot,B., et al. (2018) Public health surveillance of multidrug-resistant clones of Neisseria gonorrhoeae in Europe: a genomic survey. Lancet Infect. Dis., 18, 758–768.
```

The assemblies from the study were first annotated using Prokka

```
mkdir prokka_output
for fasta in ./genomes/*.fasta
do
prefix=$(basename ${fasta} .fasta)
prokka --cpus 5 --genus Neisseria --usegenus --outdir ./prokka_output/${prefix} --prefix $prefix $fasta
done
```

We can now run Panaroo, setting the option to build a multiple sequence alignment of the core genes using MAFFT.

```
mkdir panaroo_output
panaroo -i ./prokka_output/*/*.gff -o panaroo_output -t 24 --verbose -a core
```

In order to control for population structure in our association analyses we build a phylogeny from the core gene alignment using Iqtree. 

```
cd panaroo_output
iqtree -s core_gene_alignment.aln -pre core_tree -nt 24 -fast -m GTR
cd ..
```

We are now ready to run Pyseer. We first look at association between antibiotic resistance and gene presence/absence.

```
python ~/pyseer/scripts/phylogeny_distance.py --lmm ./panaroo_output/core_tree.treefile > pyseer_out/phylogeny_K.tsv

for anti in AZM     CRO     CFM     CIP     PEN     SMX     TET
do
python ~/pyseer/pyseer-runner.py --lmm --phenotypes ./metadata/pathogenwatch-neigo-eurogasp2013-amr-profile-1.tab --pres ./panaroo_output/gene_presence_absence.Rtab --similarity ./pyseer_out/phylogeny_K.tsv --phenotype-column $anti --output-patterns ./pyseer_out/gene_patterns_${anti}.txt > ./pyseer_out/${anti}_gwas.txt
done
```

We also have to count the number of patterns in order to control for multiple testing.

```
python ~/pyseer/scripts/count_patterns.py ./pyseer_out/gene_patterns_AZM.txt
```

This results in:

```
Patterns:       549
Threshold:      9.11E-05
```

We can also run Pyseer on the presence/absence matrix of sructural variations output by Panaroo.

```
for anti in AZM     CRO     CFM     CIP     PEN     SMX     TET
do
python ~/pyseer/pyseer-runner.py --lmm --phenotypes ./metadata/pathogenwatch-neigo-eurogasp2013-amr-profile-1.tab --pres ./panaroo_output/struct_presence_absence.Rtab --similarity ./pyseer_out/phylogeny_K.tsv --phenotype-column $anti --output-patterns ./pyseer_out/struct_patterns_${anti}.txt > ./pyseer_out/${anti}_struct_was.txt
done
```

again counting the number of patterns for false discovery analysis

```
python ~/pyseer/scripts/count_patterns.py ./pyseer_out/struct_patterns_AZM.txt
```

which gives

```
Patterns:       2779
Threshold:      1.80E-05
```

We now have everything we need to start looking at some results!

```{r}
antibiotics <- c('AZM', 'CRO', 'CFM', 'CIP', 'PEN', 'SMX', 'TET')

gono_gwas <- do.call(rbind, map(antibiotics, function(ant){
  tbl <- fread(paste(c("./gono_harris_gwas/",ant,"_gwas.txt"), collapse = ""), data.table = FALSE)
  tbl$antibiotic <- ant
  return(tbl)
}))

gono_gwas <- gono_gwas[order(gono_gwas$`lrt-pvalue`),]
gono_gwas <- gono_gwas[!grepl("bad-chisq", gono_gwas$notes),]

# threshold form running count_patterns in pyseer
sig_threshold <- 0.05/(549*length(antibiotics))

sum(gono_gwas$`lrt-pvalue`<sig_threshold)
sig_hits <- gono_gwas[gono_gwas$`lrt-pvalue`<sig_threshold,]
```

We can now add some annotation information to make things a bit easier to interpret.

```{r}
full_pa_tbl <- fread("./gono_harris_gwas/gene_presence_absence_roary.csv", data.table = FALSE)
results <- merge(sig_hits, full_pa_tbl, by.x = "variant", by.y = "Gene", all.x = TRUE)
results <- results[order(results$`lrt-pvalue`),]

write.csv(results[,c("variant", "lrt-pvalue", "beta", "beta-std-err", "Annotation")],
          file = "./gono_harris_gwas/gono_sig_hits_w_annotation.csv", quote = FALSE)
results[,c("variant", "lrt-pvalue", "beta", "beta-std-err","antibiotic", "Annotation")]

top_anno_gwas_hits <- results[results$Annotation!="hypothetical protein",c("variant", "lrt-pvalue", "beta", "beta-std-err", "antibiotic","Annotation")]

# we now output the results to csv. This can be input to cytoscape along with the final graph from Panaroo for visualisation.
write.csv(top_anno_gwas_hits, file="./gono_harris_gwas/gono_gwas_annotated.csv", quote = FALSE)
```

now we can also look at possible rearrangements

```{r}
gono_struct_was <- do.call(rbind, map(antibiotics, function(ant){
  tbl <- fread(paste(c("./gono_harris_gwas/",ant,"_struct_was.txt"), collapse = ""), data.table = FALSE)
  tbl$antibiotic <- ant
  return(tbl)
}))

gono_struct_was <- gono_struct_was[order(gono_struct_was$`lrt-pvalue`),]
gono_struct_was <- gono_struct_was[!grepl("bad-chisq", gono_struct_was$notes),]

# threshold form running count_patterns in pyseer
sig_threshold <- 0.05/(2779*length(antibiotics))

sum(gono_struct_was$`lrt-pvalue`<sig_threshold)
top_results <- gono_struct_was[gono_struct_was$`lrt-pvalue`<sig_threshold,]
top_results <- top_results[!duplicated(top_results$variant),]

write.csv(top_results, file="./gono_harris_gwas/gono_struct_was_annotated.csv", quote = FALSE)

# we can also create results file for input to cytoscape
cyto_output <- map_dfr(1:nrow(top_results), ~ data.frame(gene=unlist(str_split(top_results[.x,1], "-")),
                                          svariant=rep(top_results[.x,1], 3),
                                          lrtpvalue=rep(top_results[.x,4], 3),
                                          beta=rep(top_results[.x,5], 3), 
                                          stringsAsFactors = FALSE))
write.csv(cyto_output, file="./gono_harris_gwas/gono_struct_was_cytoscape_output.csv", quote = FALSE, row.names = FALSE)
```
