---
title: "Analysis of Gene Presence/Absence in the Pneumoccocal Global Sequenceing Project"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

##Load Libraries and Download Data

```{r, warning=FALSE, message=FALSE}
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(piggyback)
```

To demonstrate Panaroo's ability to scale to large datasets and to infer pangenome evolutionary paramters we investigated a subsection of the large Global Pneumococcal Sequenceing (GPS) project.

**Gladstone,R.A., Lo,S.W., Lees,J.A., Croucher,N.J., van Tonder,A.J., Corander,J., Page,A.J., Marttinen,P., Bentley,L.J., Ochoa,T.J., et al. (2019) International genomic definition of pneumococcal lineages, to contextualise disease, antibiotic resistance and vaccine impact. EBioMedicine, 43, 338–346.**

We only considered isolates from the major sequencing clusters (GPSC's) for which a reliable dated tree could be inferred. The dated phylogenies were obtained from **Gladstone,R.A., et al. (2019) in prep**.

##Run Panaroo

Panaroo was first run seperately on each of the GPSCs both in the strict and sensitive modes.

```
for folder in *
do

cd $folder

mkdir panaroo_out
panaroo -i *.gff -o panaroo_out -t 10 --verbose

mkdir panaroo_sensitive_out
panaroo -i *.gff -o panaroo_out -t 10 --verbose --mode relaxed

cd ..

done
```

The results of these analyses was then merged using the panaroo-merge command

```
mkdir panaroo_merged_out
panaroo-merge -d [0-9]* -o panaroo_merged_out -t 24 --verbose
```

We next used one of the Panaroo post processing functions to infer the paramters of the Infinitely Many Genes (IMG) model as described in **Collins,R.E. and Higgs,P.G. (2012) Testing the Infinitely Many Genes Model for the Evolution of the Bacterial Core Genome and Pangenome. Mol. Biol. Evol., 29, 3413–3425.** and **Baumdicker,F. and Pfaffelhuber,P. (2014) The infinitely many genes model with horizontal gene transfer. Electron. J. Probab., 19.**

```
for folder in [0-9]*
do
cd ${folder}/panaroo_out
mkdir img_model
cp ../../bactdater_outputs/GPSC${folder}_BD.tre ./
python ~/panaroo/panaroo-estimate-img.py --pa gene_presence_absence.Rtab -o img_model/ --tree GPSC${folder}_BD.tre -D 2
cd ../..
done
```

We also inferred the parameters of the Finitely Many Genes model as described in **Zamani-Dahaj,S.A., Okasha,M., Kosakowski,J. and Higgs,P.G. (2016) Estimating the Frequency of Horizontal Gene Transfer Using Phylogenetic Models of Gene Gain and Loss. Mol. Biol. Evol., 33, 1843–1857.**

```
for folder in [0-9]*
do
cd ${folder}/panaroo_out
mkdir fmg_model
python ~/panaroo/panaroo-estimate-fmg.py --model FMG --tree GPSC${folder}_BD.tre --pa gene_presence_absence_renamed.Rtab -o fmg_model/${folder}_fmg_results.txt -t 5 --nboot 100 --verbose
cd ../..
done
```
