---
title: "Comparison of Methods on a Tuberculosis Outbreak in London "
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

##Load Libraries and Download Data

```{r, warning=FALSE, message=FALSE}
library(ggplot2)
library(data.table)
library(purrr)
library(RColorBrewer)
library(piggyback)
```

```{r, eval=FALSE}
pb_download("klebsiella_pa_files.zip", 
            repo = "gtonkinhill/panaroo_manuscript", 
            dest = "./")
unzip("klebsiella_pa_files.zip")
```

To investigate the ability of different methods to accurately infer the pangenome from real data we investigated a large Tuberculosis outbreak in London:

>Casali,N., Broda,A., Harris,S.R., Parkhill,J., Brown,T. and Drobniewski,F. (2016) Whole Genome Sequence Analysis of a Large Isoniazid-Resistant Tuberculosis Outbreak in London: A Retrospective Observational Study. PLoS Med., 13, e1002137.

This provided a suitable control as *M. tuberculosis* is not thought to have a pangenome. The low mutation rate of *M. tuberculosis* (0.5 snps/year) also led to only one of these isolates having diverged by more than 5 SNPs. Thus we would not expect variation in the accessory genome and would expect nearly all genes to be 'core' genes.

##Running different pangenome tools

**Roary**

```
roary -f roary_out -p 20 ./*.gff
```

**PIRATE**

```
mkdir pirate_out
~/panaroo/scripts/run_pirate.py -i ./*.gff -o pirate_out/ -t 20 
```

This is the equivalent to running PIRATE as 

```
cp ./*.gff ./pirate_out/
PIRATE -i pirate_out -o pirate_out -t 20
```

**PanX**

```
mkdir panx_out
~/panaroo/scripts/run_panx.py -i ./*.gbk -o panx_out/ -t 20 
```

This is the equivalent to running PanX as 

```
cp ./*.gbk ./panx_out/
panX.py -fn panx_out -sl panx_run -t 20
```

**Panaroo**

```
mkdir panaroo_out
panaroo -i ./*.gff -o panaroo_out/ -t 20
```

##Loading Results

```{r}
panaroo_pa <- fread("./tb_outbreak_pa_files/panaroo_gene_presence_absence.csv", data.table = FALSE)
panaroo_pa <- as.matrix(panaroo_pa[,15:ncol(panaroo_pa)])
roary_pa <- fread("./tb_outbreak_pa_files/roary_gene_presence_absence.csv", data.table = FALSE)
roary_pa <- as.matrix(roary_pa[,15:ncol(roary_pa)])
panx_pa <- fread("./tb_outbreak_pa_files/panx_gene_presence_absence.csv", data.table = FALSE)
panx_pa <- as.matrix(panx_pa)
pirate_pa <- fread("./tb_outbreak_pa_files/PIRATE.gene_families.ordered.tsv", data.table = FALSE)
pirate_pa <- as.matrix(pirate_pa[,23:ncol(pirate_pa)])

pa_matrices <- list(panaroo_pa, roary_pa, panx_pa, pirate_pa)
names(pa_matrices) <- c("Panaroo", "Roary", "PanX", "PIRATE")
```

We can now look at the distribution of the pangenome, core and accessory sizes inferred by the different methods.

```{r}
calc_pan_sizes <- function(pa, core_fraction=0.98){
  n_isolates <- ncol(pa)
  n_core <- sum(rowSums(pa!="")>=(core_fraction*n_isolates))
  sizes <- c(nrow(pa), n_core, nrow(pa)-n_core)
  names(sizes) <- c("pangenome", "core", "accessory")
  return(sizes)
}
```

First a plot with 'core' defined as present in at least 99\% of samples. This is the default definition in Roary.

```{r}
result_99_df <- imap_dfr(pa_matrices, ~ data.frame(method=.y, 
                                                   data.frame(t(calc_pan_sizes(.x, core_fraction = 0.99))), 
                                                   stringsAsFactors = FALSE))
result_99_df <- data.table::melt(result_99_df, id.vars="method", variable.name="category", value.name="count")

ggplot(result_99_df, aes(x=method, y=count, fill=method)) + geom_col() +
  facet_wrap(~category, ncol = 1, scales = "free_y") +
  theme_bw(base_size = 16) +
  scale_fill_brewer(palette="Set1")
```

Now a plot with 'core' defined as present in at least 98\% of samples.

```{r}
result_98_df <- imap_dfr(pa_matrices, ~ data.frame(method=.y, 
                                                   data.frame(t(calc_pan_sizes(.x, core_fraction = 0.98))), 
                                                   stringsAsFactors = FALSE))
result_98_df <- data.table::melt(result_98_df, id.vars="method", variable.name="category", value.name="count")

ggplot(result_98_df, aes(x=method, y=count, fill=method)) + geom_col() +
  facet_wrap(~category, ncol = 1, scales = "free_y") +
  theme_bw(base_size = 16) +
  scale_fill_brewer(palette="Set1")
```

Now a plot with 'core' defined as present in at least 95\% of samples

```{r}
result_95_df <- imap_dfr(pa_matrices, ~ data.frame(method=.y, 
                                                   data.frame(t(calc_pan_sizes(.x, core_fraction = 0.95))), 
                                                   stringsAsFactors = FALSE))
result_95_df <- data.table::melt(result_95_df, id.vars="method", variable.name="category", value.name="count")

ggplot(result_95_df, aes(x=method, y=count, fill=method)) + geom_col() +
  facet_wrap(~category, ncol = 1, scales = "free_y") +
  theme_bw(base_size = 16) +
  scale_fill_brewer(palette="Set1")
```





